src/ows/src/main/java/org/geoserver/ows/XmlRequestReader.java

```diff
@@ -126,4 +130,73 @@ public int hashCode() {
     public String getServiceId() {
         return serviceId;
     }
+
+    /**
+     * Wrap request parsing failure in a SAX Document to prevent generic FileNotFound and
+     * ConnectionRefused messages.
+     *
+     * <p>This communicates that there is a problem with the XML Document while avoiding providing
+     * internal details.
+     *
+     * <p>The full stack trace is preserved to support development and debugging.
+     *
+     * @param t Parsing failure
+     * @return Clean SAXException for common parsing failures, or initial throwable
+     */
+    protected Exception cleanException(Exception t) {
+        if (t instanceof IOException) {
+            // FileNotFoundException, ConnectionException, etc...
+            // Log actual failure for debugging and troubleshooting
+            final Logger LOGGER = Logging.getLogger(this.getClass());
+            String requestFailure = "XML " + getElement().getLocalPart() + " Parsing Failure: ";
+            String cleanMessage =
+                    "Parsing failed, the xml request is most probably not compliant to "
+                            + getElement().getLocalPart()
+                            + " element";
+
+            LOGGER.fine(requestFailure + t.toString());
+
+            // Clean SAXException message, keep stacktrace history (for verbose service exception
+            // document)
+            SAXException parseException = new SAXException(cleanMessage);
+            parseException.setStackTrace(t.getStackTrace());
+            return parseException;
+        }
+        if (t instanceof SAXException) {
+            // Double check SAXException does not echo
+            return cleanSaxException((SAXException) t);
+        }
+        return t;
+    }
+
+    /**
+     * Clean the localized message, in the case where it is reproduced by SAXException default
+     * constructor.
+     *
+     * @param saxException
+     * @return saxException with nested localized message removed.
+     */
+    protected SAXException cleanSaxException(SAXException saxException) {
+        Throwable cause = saxException.getCause();
+        // We only wish to clean the generic sax exception that recycles internal cause message
+        if (saxException != null
+                && saxException.getCause() != null
+                && saxException.getClass() == SAXException.class) {
+            String saxMessage = saxException.getMessage();
+            String causeMessage = cause.getLocalizedMessage();
+            if (causeMessage != null && saxMessage.contains(causeMessage)) {
```

Comment:
Just wanted to make sure this "if" is correct.  It looks a little funny since the exception isn't cleaned unless the saxMessage contains the causeMessage.